test/e2e/restart.go

```diff
@@ -0,0 +1,367 @@
+		By("ensuring the same number of pods are running and ready after restart")
+		podCheckStart := time.Now()
+		podNamesAfter, err := waitForNPods(c, len(podNamesBefore), restartPodReadyAgainTimeout)
+		Expect(err).NotTo(HaveOccurred())
+		remaining := restartPodReadyAgainTimeout - time.Since(podCheckStart)
+		if !checkPodsRunningReady(c, podNamesAfter, remaining) {
+			Failf("At least one pod wasn't running and ready after the restart.")
+		}
+	})
+})
+
+// waitForNPods tries to list pods using c until it finds expect of them,
+// returning their names if it can do so before timeout.
+func waitForNPods(c *client.Client, expect int, timeout time.Duration) ([]string, error) {
+	// Loop until we find expect pods or timeout is passed.
+	var podList *api.PodList
+	var err error
+	found, start := false, time.Now()
+	for ; time.Since(start) < timeout; time.Sleep(poll) {
+		podList, err = listPods(c, api.NamespaceDefault, labels.Everything(), fields.Everything())
+		if err != nil {
+			Logf("Error getting pods: %v", err)
+			continue
+		}
+		if len(podList.Items) != expect {
+			err = fmt.Errorf("expected to find %d pods but found only %d",
+				expect, len(podList.Items))
+			Logf("Error getting pods: %v", err)
+			continue
+		}
+		found = true
+		break
+	}
+	podNames := make([]string, len(podList.Items))
+	if !found {
+		return podNames, fmt.Errorf("couldn't find %d pods within %v; last error: %v",
+			expect, timeout, err)
+	}
+
+	// Successfully found the desired number of pods; extract their names.
+	for i, p := range podList.Items {
+		podNames[i] = p.ObjectMeta.Name
+	}
+	return podNames, nil
+}
+
+// checkNodesReady waits up to nt for expect nodes accessed by c to be ready,
+// returning an error if this doesn't happen in time. It returns the names of
+// nodes it finds.
+func checkNodesReady(c *client.Client, nt time.Duration, expect int) ([]string, error) {
+	// First, keep getting all of the nodes until we get the number we expect.
+	var nodeList *api.NodeList
+	var err error
+	found, start := false, time.Now()
+	for ; time.Since(start) < nt; time.Sleep(poll) {
+		// Even though listNodes(...) has its own retries, a rolling-update
+		// (GCE/GKE implementation of restart) can complete before the apiserver
+		// knows about all of the nodes. Thus, we retry the list nodes call
+		// until we get the expected number of nodes.
+		nodeList, err = listNodes(c, labels.Everything(), fields.Everything())
+		if err != nil {
+			Logf("Error getting nodes: %v", err)
+			continue
+		}
+		if len(nodeList.Items) != expect {
+			err = fmt.Errorf("expected to find %d nodes but found only %d",
+				expect, len(nodeList.Items))
+			Logf("Error getting nodes: %v", err)
+			continue
+		}
+		found = true
+		break
+	}
+	nodeNames := make([]string, len(nodeList.Items))
+	if !found {
+		return nodeNames, fmt.Errorf("couldn't find %d nodes within %v; last error: %v",
+			expect, nt, err)
+	}
+
+	// Next, ensure in parallel that all the nodes are ready. We subtract the
```

Comment:
Ugh, yeah, I spent a while trying to figure out the best way for this to be factored... all this really does is run _n_ `waitForNodeToBeReady(...)` in parallel and check their result. However, because of our Go 1.3 support and my extra-safe "waiting until all nodes are done before failing," this a handful extra lines of code. I din't factor it out more because I wouldn't reuse the function if I did.

Please let me know your final decision after my whining above :-)